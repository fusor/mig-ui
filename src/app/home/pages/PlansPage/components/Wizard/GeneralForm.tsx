import React, { useEffect, useRef, useState } from 'react';
import { FormikProps } from 'formik';
import { IFormValues, IOtherProps } from './WizardContainer';
import { Form, FormGroup, Grid, GridItem, TextInput, Title, Button } from '@patternfly/react-core';
import { CheckIcon } from '@patternfly/react-icons';
import spacing from '@patternfly/react-styles/css/utilities/Spacing/spacing';
import SimpleSelect, { OptionWithValue } from '../../../../../common/components/SimpleSelect';
import AddEditTokenModal from '../../../../../common/components/AddEditTokenModal';
import { useOpenModal } from '../../../../duck/hooks';
import IconWithText from '../../../../../common/components/IconWithText';
import TokenSelect from './TokenSelect';
const styles = require('./GeneralForm.module');

interface IGeneralFormProps
  extends Pick<IOtherProps, 'clusterList' | 'storageList' | 'tokenList' | 'isEdit'>,
    Pick<
      FormikProps<IFormValues>,
      | 'handleBlur'
      | 'handleChange'
      | 'setFieldTouched'
      | 'setFieldValue'
      | 'values'
      | 'touched'
      | 'errors'
    > {}

const GeneralForm: React.FunctionComponent<IGeneralFormProps> = ({
  clusterList,
  storageList,
  tokenList,
  errors,
  handleBlur,
  handleChange,
  isEdit,
  setFieldTouched,
  setFieldValue,
  touched,
  values,
}: IGeneralFormProps) => {
  const planNameInputRef = useRef(null);
  useEffect(() => {
    planNameInputRef.current.focus();
  }, []);

  const onHandleChange = (val, e) => handleChange(e);

  let srcClusterOptions: string[] = ['No valid clusters found'];
  let targetClusterOptions: string[] = [];
  let storageOptions: string[] = ['No valid storage found'];

  if (clusterList.length) {
    srcClusterOptions = clusterList
      .filter(
        (cluster) =>
          cluster.MigCluster.metadata.name !== values.targetCluster &&
          cluster.ClusterStatus.hasReadyCondition
      )
      .map((cluster) => cluster.MigCluster.metadata.name);

    targetClusterOptions = clusterList
      .filter(
        (cluster) =>
          cluster.MigCluster.metadata.name !== values.sourceCluster &&
          cluster.ClusterStatus.hasReadyCondition
      )
      .map((cluster) => cluster.MigCluster.metadata.name);
  }

  if (storageList.length) {
    storageOptions = storageList
      .filter((storage) => storage.StorageStatus.hasReadyCondition)
      .map((storage) => storage.MigStorage.metadata.name);
  }

  const handleStorageChange = (value) => {
    const matchingStorage = storageList.find((c) => c.MigStorage.metadata.name === value);
    if (matchingStorage) {
      setFieldValue('selectedStorage', value);
      setFieldTouched('selectedStorage');
    }
  };

  const handleSourceChange = (value) => {
    const matchingCluster = clusterList.find((c) => c.MigCluster.metadata.name === value);
    if (matchingCluster) {
      setFieldValue('sourceCluster', value);
      setFieldValue('selectedNamespaces', []);
      setFieldTouched('sourceCluster');
      setFieldValue('sourceToken', null);
    }
  };

  const handleTargetChange = (value) => {
    const matchingCluster = clusterList.find((c) => c.MigCluster.metadata.name === value);
    if (matchingCluster) {
      setFieldValue('targetCluster', value);
      setFieldTouched('targetCluster');
      setFieldValue('targetToken', null);
    }
  };

  return (
    <Form>
      <Title size="md" className={styles.fieldGridTitle}>
        Give your plan a name
      </Title>

      <Grid md={6} gutter="md" className={spacing.mbMd}>
        <GridItem>
          <FormGroup
            label="Plan name"
            isRequired
            fieldId="planName"
            helperTextInvalid={touched.planName && errors.planName}
            isValid={!(touched.planName && errors.planName)}
          >
            <TextInput
              ref={planNameInputRef}
              onChange={(val, e) => onHandleChange(val, e)}
              onInput={() => setFieldTouched('planName', true, true)}
              onBlur={handleBlur}
              value={values.planName}
              name="planName"
              type="text"
              isValid={!errors.planName && touched.planName}
              id="planName"
              isDisabled={isEdit}
            />
          </FormGroup>
        </GridItem>
      </Grid>

      <Title size="md" className={styles.fieldGridTitle}>
        Select source and target clusters
      </Title>

      <Grid md={6} gutter="md" className={spacing.mbMd}>
        <GridItem>
          <FormGroup
            label="Source cluster"
            isRequired
            fieldId="sourceCluster"
            helperTextInvalid={touched.sourceCluster && errors.sourceCluster}
            isValid={!(touched.sourceCluster && errors.sourceCluster)}
          >
            <SimpleSelect
              id="sourceCluster"
              onChange={handleSourceChange}
              options={srcClusterOptions}
              value={values.sourceCluster}
              placeholderText="Select source..."
            />
          </FormGroup>
          <TokenSelect
            fieldId="sourceToken"
            tokenList={[
              // NATODO just pass in tokenList prop
              {
                MigToken: {
                  apiVersion: '1.0',
                  kind: 'MigToken',
                  metadata: {
                    name: 'My Token',
                  },
                  spec: {
                    migClusterRef: { name: 'host', namespace: 'some-namespace' },
                    secretRef: { name: 'secret-name', namespace: 'some-namespace' },
                  },
                  status: {
                    observedDigest: 'foo',
                    type: 'OAuth',
                    expiresAt: '2020-06-08T20:55:53.825Z',
                  },
                },
              },
            ]}
            clusterName={values.sourceCluster}
            value={values.sourceToken}
            onChange={(tokenName) => {
              setFieldTouched('sourceToken');
              setFieldValue('sourceToken', tokenName);
            }}
            touched={touched.sourceToken}
            error={errors.sourceToken}
            expiringSoonMessage="The users's selected token on cluster source will expire soon."
            expiredMessage="The user's selected token on cluster source is expired."
          />
        </GridItem>

        <GridItem>
          <FormGroup
            label="Target cluster"
            isRequired
            fieldId="targetCluster"
            helperTextInvalid={touched.targetCluster && errors.targetCluster}
            isValid={!(touched.targetCluster && errors.targetCluster)}
          >
            <SimpleSelect
              id="targetCluster"
              onChange={handleTargetChange}
              options={targetClusterOptions}
              value={values.targetCluster}
              placeholderText="Select target..."
            />
          </FormGroup>
          <TokenSelect
            fieldId="targetToken"
            tokenList={tokenList}
            clusterName={values.targetCluster}
            value={values.targetToken}
            onChange={(tokenName) => {
              setFieldTouched('targetToken');
              setFieldValue('targetToken', tokenName);
            }}
            touched={touched.targetToken}
            error={errors.targetToken}
            expiringSoonMessage="The users's selected token on cluster target will expire soon."
            expiredMessage="The user's selected token on cluster target is expired."
          />
        </GridItem>
      </Grid>

      <Title size="md" className={styles.fieldGridTitle}>
        Select a replication repository
      </Title>

      <Grid md={6} gutter="md">
        <GridItem>
          <FormGroup
            label="Repository"
            isRequired
            fieldId="selectedStorage"
            helperTextInvalid={touched.selectedStorage && errors.selectedStorage}
            isValid={!(touched.selectedStorage && errors.selectedStorage)}
          >
            <SimpleSelect
              id="selectedStorage"
              onChange={handleStorageChange}
              options={storageOptions}
              value={values.selectedStorage}
              placeholderText="Select repository..."
            />
          </FormGroup>
        </GridItem>
      </Grid>
    </Form>
  );
};

export default GeneralForm;
