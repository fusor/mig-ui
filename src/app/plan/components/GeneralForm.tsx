import React from "react";
import { Route, Redirect, RouteComponentProps } from "react-router-dom";
import { Flex, Box, Text } from "@rebass/emotion";
import {
  Button,
  TextInput,
  TextContent,
  TextList,
  TextListItem,
  TextArea
} from "@patternfly/react-core";

interface IProps {
  component: React.ReactNode;
}

const GeneralForm: React.SFC<IProps & RouteComponentProps> = ({
  handleChange,
  handleBlur,
  values,
  errors,
  touched,
  setFieldTouched,
  ...rest
}) => (
  <React.Fragment>
    <Box>
      <TextContent>
        <TextList component="dl">
          <TextListItem component="dt">Plan Name</TextListItem>
          <input
            onChange={handleChange}
            onInput={() => setFieldTouched("planName", true, true)}
            onBlur={handleBlur}
            value={values.planName}
            name="planName"
            type="text"
          />
          {errors.planName && touched.planName && (
            <div id="feedback">{errors.planName}</div>
          )}
        </TextList>
      </TextContent>
    </Box>
  </React.Fragment>
);

export default GeneralForm;
