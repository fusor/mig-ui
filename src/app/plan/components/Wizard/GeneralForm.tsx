import React from 'react';
import { RouteComponentProps } from 'react-router-dom';
import { Box, Text } from '@rebass/emotion';
import {
  Form,
  FormGroup,
  Grid,
  GridItem,
  TextInput
} from '@patternfly/react-core';
import FormErrorDiv from '../../../common/components/FormErrorDiv';
interface IProps {
  component: React.ReactNode;
}

const GeneralForm: React.SFC<IProps & RouteComponentProps> = ({
  handleChange,
  handleBlur,
  values,
  errors,
  touched,
  setFieldTouched,
  ...rest
}) => {
  const onHandleChange = (val, e) => {
    handleChange(e);
  };

  return (
    <Form>
      <Grid lg={10} xl={6}>
        <GridItem>
          <FormGroup
            label="Plan Name"
            isRequired
            fieldId="planName"
          >
            <TextInput
              onChange={(val, e) => onHandleChange(val, e)}
              onInput={() => setFieldTouched('planName', true, true)}
              onBlur={handleBlur}
              value={values.planName}
              name="planName"
              type="text"
              isValid={!errors.planName && touched.planName}
              id="planName"
            />
          </FormGroup>
        </GridItem>
      </Grid>
    </Form>
  );
};

export default GeneralForm;
