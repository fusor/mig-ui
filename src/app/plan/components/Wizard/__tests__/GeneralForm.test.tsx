import React from 'react';
import ReactDOM from 'react-dom';
import { render, cleanup, screen, fireEvent } from '@testing-library/react';
import GeneralForm, { IGeneralFormProps } from '../GeneralForm';
import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';
import { planValue } from './planHelper';

const defaultProps: IGeneralFormProps = {
  handleBlur() {
    return;
  },
  handleChange() {
    return;
  },
  setFieldTouched() {
    return;
  },
  touched: {
    planName: true,
  },
  errors: '',
  isEdit: true,
  values: planValue,
};

function renderGeneralForm(props: Partial<IGeneralFormProps> = {}) {
  return render(<GeneralForm {...defaultProps} {...props} />);
}

// afterEach(cleanup);

describe('<GeneralForm />', () => {
  it('renders component', () => {
    const div = document.createElement('div');
    ReactDOM.render(<GeneralForm {...defaultProps}></GeneralForm>, div);
  });

  it('has a plan name', async () => {
    renderGeneralForm();
    const planForm = await screen.findByTestId('plan-form');

    expect(planForm).toHaveFormValues({
      planName: 'plan-name-test',
    });
  });

  it('renders snapshot', async () => {
    const tree = renderer.create(<GeneralForm {...defaultProps}></GeneralForm>).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
