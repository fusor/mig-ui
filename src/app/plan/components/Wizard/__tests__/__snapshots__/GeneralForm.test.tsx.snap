// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GeneralForm /> render snapshot 1`] = `
<form
  className="pf-c-form"
  data-testid="plan-form"
  noValidate={true}
>
  <div
    className="pf-l-grid pf-m-all-10-col-on-lg pf-m-all-6-col-on-xl"
  >
    <div
      className="pf-l-grid__item"
    >
      <div
        className="pf-c-form__group"
      >
        <label
          className="pf-c-form__label"
          htmlFor="planName"
        >
          <span
            className="pf-c-form__label-text"
          >
            Plan Name
          </span>
          <span
            aria-hidden="true"
            className="pf-c-form__label-required"
          >
            *
          </span>
        </label>
        <input
          aria-invalid={false}
          aria-label={null}
          className="pf-c-form-control"
          disabled={true}
          id="planName"
          name="planName"
          onBlur={[Function]}
          onChange={[Function]}
          onInput={[Function]}
          readOnly={false}
          required={false}
          type="text"
          value="plan-name-test"
        />
        
      </div>
    </div>
  </div>
</form>
`;
